File: app/src/main/java/com/bitflaker/lucidsourcekit/data/FrequencyList.java
package com.bitflaker.lucidsourcekit.data;

import java.util.ArrayList;
import java.util.List;

public class FrequencyList {
    private final List<FrequencyData> frequencyData = new ArrayList<>();

    public FrequencyData get(int i){
        return frequencyData.get(i);
    }

    public void add(FrequencyData data) {
        int maxDuration = 2000;
        if(data.getDuration() <= maxDuration){
            frequencyData.add(data);
        }
        else {
            float from = data.getFrequency();
            float to = Float.isNaN(data.getFrequencyTo()) ? from : data.getFrequencyTo();
            float diff = to - from;
            int count = (int)Math.ceil(data.getDuration() / (float) maxDuration);
            float step = diff / count;
            float durInc = data.getDuration() / (float) count;
            float currFreq = data.getFrequency();
            for (int i = 0; i < count; i++) {
                frequencyData.add(new FrequencyData(currFreq, currFreq + step, durInc));
                currFreq += step;
            }
        }
    }

    public void remove(int i){
        frequencyData.remove(i);
    }

    public void remove(FrequencyData data){
        frequencyData.remove(data);
    }

    public float getDuration() {
        return (float) frequencyData.stream().mapToDouble(FrequencyData::getDuration).sum();
    }

    public int size(){
        return frequencyData.size();
    }

    public float getDurationUntilAfter(int i) {
        return getDurationUntil(i) + frequencyData.get(i).getDuration();
    }

    public float getDurationUntil(int i) {
        float duration = 0.0f;
        for (int j = 0; j < i; j++){
            duration += frequencyData.get(j).getDuration();
        }
        return duration;
    }

    public double getFrequencyAtDuration(double duration) {
        double durationCounter = 0.0;
        for (int i = 0; i < frequencyData.size(); i++){
            FrequencyData current = frequencyData.get(i);
            durationCounter += current.getDuration();
            if(durationCounter >= duration) {
                if(Float.isNaN(current.getFrequencyTo())){
                    return current.getFrequency();
                }
                else{
                    double k = (current.getFrequencyTo() - current.getFrequency()) / (double)current.getDuration();
                    return k * (duration - (durationCounter - current.getDuration())) + current.getFrequency();
                }
            }
        }
        return -1;
    }
}

File: app/src/main/java/com/bitflaker/lucidsourcekit/main/binauralbeats/player/FrequencyList.kt
package com.bitflaker.lucidsourcekit.main.binauralbeats.player

import kotlin.math.ceil

class FrequencyList {
    private val frequencyData = ArrayList<FrequencyData>()

    operator fun get(i: Int): FrequencyData {
        return frequencyData[i]
    }

    fun add(data: FrequencyData) {
        val maxDuration = 2000
        if (data.duration <= maxDuration) {
            frequencyData.add(data)
        } else {
            val from = data.frequency
            val to = if (data.getFrequencyTo().isNaN()) from else data.getFrequencyTo()
            val diff = to - from
            val count = ceil((data.duration / maxDuration.toFloat()).toDouble()).toInt()
            val step = diff / count
            val durInc = data.duration / count.toFloat()
            var currFreq = data.frequency
            for (i in 0..<count) {
                frequencyData.add(FrequencyData(currFreq, currFreq + step, durInc))
                currFreq += step
            }
        }
    }

    fun remove(i: Int) {
        frequencyData.removeAt(i)
    }

    fun remove(data: FrequencyData) {
        frequencyData.remove(data)
    }

    val duration: Float
        get() = frequencyData.stream().mapToDouble { it.duration.toDouble() }.sum().toFloat()

    fun size(): Int {
        return frequencyData.size
    }

    fun getDurationUntilAfter(i: Int): Float {
        return getDurationUntil(i) + frequencyData[i].duration
    }

    fun getDurationUntil(i: Int): Float {
        var duration = 0.0f
        for (j in 0..<i) {
            duration += frequencyData[j].duration
        }
        return duration
    }

    fun getFrequencyAtDuration(duration: Double): Double {
        var durationCounter = 0.0
        for (i in frequencyData.indices) {
            val current = frequencyData[i]
            durationCounter += current.duration.toDouble()
            if (durationCounter >= duration) {
                if (current.getFrequencyTo().isNaN()) {
                    return current.frequency.toDouble()
                }
                val k = (current.getFrequencyTo() - current.frequency) / current.duration.toDouble()
                return k * (duration - (durationCounter - current.duration)) + current.frequency
            }
        }
        return -1.0
    }
}
